### PR TITLE
[multistage] Support IN and NOT-IN Clauses

### PR DESCRIPTION
**Pull request by ankitsultana**
Related to #9345 

Query to verify:
```
SELECT COUNT(*)
        FROM baseballStats_OFFLINE
        WHERE teamID IN ('BL2', 'FL1')
```

cc: @walterddr

### CODE REVIEW

**Review by walterddr — COMMENTED** [review]
could you add some test to validate this?

**Comment by walterddr** [comment]
this should be included in https://github.com/apache/pinot/pull/9375

**Comment by ankitsultana** [comment]
@walterddr : I was working on the IN/NOT-IN/Range support. Does #9375 take care of that as well?

**Comment by walterddr** [comment]
> @walterddr : I was working on the IN/NOT-IN/Range support. Does #9375 take care of that as well?

ah. no that only takes care of the `CHAR/VARCHAR` codepath to decode NlsString into regular String

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java [hunk 1]
```diff
@@ -106,6 +113,33 @@ static FieldSpec.DataType toDataType(RelDataType type) {
     }
   }
 
+  // POC Code. Will clean-up later.
+  static RexExpression handleSearch(RexCall rexCall) {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java:117]: we can create a new util class call `RexExpressionUtils` to handle special SqlKinds

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java [hunk 2]
```diff
@@ -146,7 +180,7 @@ class Literal implements RexExpression {
     public Literal() {
     }
 
-    public Literal(FieldSpec.DataType dataType, SqlTypeName sqlTypeName, @Nullable Object value) {
+    public Literal(FieldSpec.DataType dataType, @Nullable Object value) {
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java:183]: good catch

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 3]
```diff
@@ -100,6 +100,12 @@ private Object[][] provideTestSqlAndRowCount() {
         new Object[]{"SELECT a.col1, a.ts, b.col2, b.col3 FROM a JOIN b ON a.col1 = b.col2 "
             + " WHERE a.col3 >= 0 AND a.col2 = 'alice' AND b.col3 >= 0", 3},
 
+        // Join query with IN and Not-IN clause. Table A's side of join will return 9 rows and Table B's side will
+        // return 2 rows. Join will be only on col1=bar and since A will return 3 rows with that value and B will return
+        // 1 row, the final output will have 3 rows.
+        new Object[]{"SELECT a.col1, b.col2 FROM a JOIN b ON a.col1 = b.col1 "
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:106]: can we add more test. `IN` and `NOT IN` with 0, 1, and more than 1 values

**Review by walterddr — COMMENTED** [review]
looks good to me overall with the approach. left a few comments. plz kindly take a look

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java (hunk 3 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:106]: also add test for implicit range search, such as 
- `WHERE (a.col3 > 0 AND a.col3 < 10) OR a.col3 >= 30`
- `WHERE a.col2 = 'foo' OR a.col2 = 'bar'`

Comment (ankitsultana) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:106]: I can add the range search test in the range PR that I'll raise later this week?

This PR only focuses on IN/Not-In. Ranges won't work yet.

Comment (ankitsultana) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:115]: Updated. Query fails parsing if there are no arguments to the IN clause, i.e. `col1 IN ()` fails

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java [hunk 4]
```diff
@@ -52,30 +51,32 @@ static RexExpression toRexExpression(RexNode rexNode) {
...
         // CAST is being rewritten into "rexCall.CAST<targetType>(inputValue)",
         //   - e.g. result type has already been converted into the CAST RexCall, so we assert single operand.
+        operands = rexCall.getOperands().stream().map(RexExpression::toRexExpression).collect(Collectors.toList());
         Preconditions.checkState(operands.size() == 1, "CAST takes exactly 2 arguments");
         RelDataType castType = rexCall.getType();
         // add the 2nd argument as the source type info.
-        operands.add(new Literal(FieldSpec.DataType.STRING, rexCall.getOperands().get(0).getType().getSqlTypeName(),
+        operands.add(new Literal(FieldSpec.DataType.STRING,
             toPinotDataType(rexCall.getOperands().get(0).getType()).name()));
         return new RexExpression.FunctionCall(rexCall.getKind(), toDataType(rexCall.getType()), "CAST", operands);
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/RexExpression.java:75]: can you also refactor this into RexExpressionUtils?

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 5]
```diff
@@ -109,6 +109,23 @@ private Object[][] provideTestSqlAndRowCount() {
...
+            + " WHERE a.col1 IN ('foo', 'bar', 'alice') AND b.col2 NOT IN ('foo', 'alice')", 3},
```
Comment (walterddr) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:116]: can you also add a test in QueryCompilationTest (QueryEnvironmentTestBase) for the parsing? just one single entry should be enough by mixing all types of predicates together

Comment (ankitsultana) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:116]: Done. Also added test for empty IN clause and range predicates to verify they are failing. The range test I'll move to testPlanWithoutException in the range PR.

**Pull request merged**